src/toil/wdl/wdl_types.py

```diff
@@ -1,41 +1,211 @@
+        """
+        if value is None:
+            # check if input is in fact an optional.
+            if self.optional:
+                return self.default_value
+            else:
+                raise WDLRuntimeError(f"Required input for '{self.name}' type not specified.")
+
+        return self._create(value)
+
+    def _create(self, value: Any) -> Any:
+        raise NotImplementedError
+
+    def __eq__(self, other):
+        # TODO: remove after refactor
+        return self.name.__eq__(other)
+
+    def __str__(self):
+        return self.name.__str__()
+
+    def __repr__(self):
+        return self.name.__repr__()
+
+
+class WDLCompoundType(WDLType, ABC):
+    """
+    Represent a WDL compound type.
+    """
     pass
 
 
-class WDLPairType(WDLType):
-    def __new__(cls, left: str, right: str):
-        return super(WDLPairType, cls).__new__(cls, 'Pair')
+class WDLStringType(WDLType):
+    """ Represent a WDL String primitive type."""
 
-    def __init__(self, left: str, right: str):
-        super(WDLPairType, self).__init__()
-        self.left = left
-        self.right = right
+    @property
+    def name(self) -> str:
+        return 'String'
+
+    @property
+    def default_value(self):
+        return ''
+
+    def _create(self, value: Any) -> Any:
+        return str(value)
+
+
+class WDLIntType(WDLType):
+    """ Represent a WDL Int primitive type."""
+
+    @property
+    def name(self) -> str:
+        return 'Int'
+
+    def _create(self, value: Any) -> Any:
+        return int(value)
+
+
+class WDLFloatType(WDLType):
+    """ Represent a WDL Float primitive type."""
+
+    @property
+    def name(self) -> str:
+        return 'Float'
+
+    def _create(self, value: Any) -> Any:
+        return float(value)
 
-    def __repr__(self):
-        return f'WDLPairType({repr(self.left)}, {repr(self.right)})'
 
+class WDLBooleanType(WDLType):
+    """ Represent a WDL Boolean primitive type."""
 
-class WDLMapType(WDLType):
-    def __new__(cls, key: str, value: str):
-        return super(WDLMapType, cls).__new__(cls, 'Map')
+    @property
+    def name(self) -> str:
+        return 'Boolean'
 
-    def __init__(self, key: str, value: str):
-        super(WDLMapType, self).__init__()
+    def _create(self, value: Any) -> Any:
+        return True if value else False
+
+
+class WDLFileType(WDLType):
+    """ Represent a WDL File primitive type."""
+
+    @property
+    def name(self) -> str:
+        return 'File'
+
+    @property
+    def default_value(self):
+        return ''
+
+    def _create(self, value: Any) -> Any:
+        if isinstance(value, tuple):
+            return value
+
+        # Unprocessed and unread files are represented as a tuple: (file_path, preserved_file_name).
+        # Unprocessed files (not imported to the fileStore) are tuples in the form of (file_path, None).
+        # Once the file is read (via readGlobalFile), it will change to the absolute path as a string.
```

Comment:
This seems a bit complicated; maybe there should be a `WDLFile` class that can represent these states and give you named methods to ask about them.

Then instead of trying to guess why we care about `isinstance(x, tuple)` and `if x[1]:` we would be able to see that we are really doing `isinstance(x, WDLFile)` and `if x.is_processed()`. Otherwise, you kind of want comments at every point of use to explain stuff, like "Since it's a tuple, it represents a file".

I'd also quibble with "processed" as the word for having saved the file to the file store; it's too generic. Maybe you can call them "written" and "read" to match the methods on the file store?

Also, what actually is the lifecycle of a file? What are all the possible states and the transitions between them? It can't be that an unprocessed file is always `(file_path, None)` if the unprocessed files that are also unread are `(file_path, preserved_file_name)`. And how are processed but not yet read files represented?

Making this into its own class instead of a particular language of tuples might make it easier to explain the underlying concepts. If you don't do that, I think you should expand this description to more fully explain what kinds of files there are, and how each is represented as a tuple, and how and when one becomes another.

Reply:
Using a `WDLFile` class would indeed be better. Thanks for the suggestion! I will update my PR.